### PR TITLE
#160609557 comment on articles

### PR DESCRIPTION
**Pull request by MuhanguziDavid**
### What does this PR do?

Allows logged in users to create comments on articles and send replies to those comments.

### Description of tasks to be completed?

Currently, a user cannot comment on an article.
This PR adds functionality to allow logged in users to comment on articles and also add replies to comments

### How should this be manually tested?

Make migrations `python manage.py makemigrations` and `python manage.py migrate`
Run the server `python manage.py runserver`

NB: Please check the background context section before proceeding

***Create a comment***
Open postman and copy this URL POST `http://127.0.0.1:8000/api/articles/article-1/comments/`.
Add a comment body in the format shown below:
```
{
	"comment": {
		"body": "this is a comment, a what, a comment, a what, a comment, oh!  a comment"
	}
}
```
A comment should be created and its details should be returned.

***Retreive comments***
 GET `http://127.0.0.1:8000/api/articles/article-1/comments/` in Postman.
Comments created should be returned.

***Edit a comment***
PUT `http://127.0.0.1:8000/api/articles/article-1/comments/1/` in postman
Add a comment body in the format shown below:
```
{
	"comment": {
		"body": "Edited comment"
	}
}
```
The comment should be edited and the edited body should be returned.

***Delete a comment***
DELETE `http://127.0.0.1:8000/api/articles/article-1/comments/1/` in postman
The comment should be deleted and you should get a message saying that the comment was deleted successfully.

***Post a thread to a comment***
POST `http://127.0.0.1:8000/api/articles/article-1/comments/1/thread/` in postman
Add a comment body in the format shown below:
```
{
	"comment": {
		"body": "This is a comment reply"
	}
}
```
The thread should be created and the thread details should be returned.

***Retrieve a thread of replies to a comment***
GET `http://127.0.0.1:8000/api/articles/article-1/comments/1/thread/` in postman
The thread of replies should be returned.

### What are the relevant pivotal tracker stories?

[#160609557](https://www.pivotaltracker.com/story/show/160609557)

### Any background context you want to add?

Ensure to login (all endpoints concerning comments are secured) and have an article created before proceeding to create a comment.
Create an article with the title "article 1" in order to proceed with manual testing of this PR.

### Screenshots
*Response from creating a comment*
post a comment

*Response from getting comments*
get comments

*Response from editing a comment*
edit comment

*Response from deleting a comment*
delete comment

*Response from creating a thread*
thread to a comment

*Response from getting a thread to a comment*
retrieve a thread

### CODE REVIEW

**Review by kakaemma — COMMENTED** [review]
Add comments to the article and a thread count to the comments.

File: authors/apps/articles/renderers.py [hunk 1]
```diff
@@ -24,3 +28,51 @@ def render(self, data, media_type=None, renderer_context=None):
...
+class CommentJSONRenderer(JSONRenderer):
+    """Renderer for the comments"""
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        """Method to convert data to a specific format"""
+        if type(data) != ReturnList:
+            errors = data.get('errors', None)
+            if errors is not None:
+                return json.dumps({
```
Comment (kakaemma) [inline authors/apps/articles/renderers.py]: Have the return statement on one line.

File: authors/apps/articles/renderers.py [hunk 2]
```diff
@@ -24,3 +28,51 @@ def render(self, data, media_type=None, renderer_context=None):
...
+                    'comment': data
+                })
+
+        if type(data) == ReturnDict:
+            return json.dumps({
```
Comment (kakaemma) [inline authors/apps/articles/renderers.py]: The previous comment applies here.

File: authors/apps/articles/renderers.py [hunk 3]
```diff
@@ -24,3 +28,51 @@ def render(self, data, media_type=None, renderer_context=None):
...
+
+class ThreadJSONRenderer(JSONRenderer):
+    """Renderer for the comments"""
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        """Method to convert data to a specific format"""
+        if type(data) != ReturnList:
+            errors = data.get('errors', None)
+            if errors is not None:
```
Comment (kakaemma) [inline authors/apps/articles/renderers.py]: Refactor this into a reusable method.

File: authors/apps/articles/tests.py [hunk 4]
```diff
@@ -318,3 +318,176 @@ def populate_impression_table():
...
+                "username": self.username,
+                "email": self.email,
+                "password": self.password
+            }
+        }
+        self.response = self.client.post(
+            '/api/users/', self.user_data, format="json"
+        )
+        self.client.credentials(
+            HTTP_AUTHORIZATION='Token ' + self.response.data['auth_token']
```
Comment (kakaemma) [inline authors/apps/articles/tests.py]: use ` .formart()` instead of `+`

File: authors/apps/articles/tests.py [hunk 5]
```diff
@@ -318,3 +318,176 @@ def populate_impression_table():
...
+                self.create_article.data['article']['slug'],
+                self.create_comment.data['id']
+            ),
+            self.thread,
+            format="json"
+        )
+
+    def test_comment_creation(self):
+        self.assertEqual(self.create_comment.status_code,
+                         status.HTTP_201_CREATED)
```
Comment (kakaemma) [inline authors/apps/articles/tests.py]: Use the agreed upon code alignment convention.

File: authors/apps/articles/tests.py [hunk 6]
```diff
@@ -318,3 +318,176 @@ def populate_impression_table():
...
+        )
+        self.assertEqual(result.status_code, status.HTTP_200_OK)
+
+    def test_comment_update(self):
+        result = self.client.put(
+            '/api/articles/{}/comments/{}/'.format(
+                self.create_article.data['article']['slug'],
+                self.create_comment.data['id']
+            ),
+            self.comment,
```
Comment (kakaemma) [inline authors/apps/articles/tests.py]: Put `self.comment` and `format` on the same line

File: authors/apps/articles/tests.py [hunk 7]
```diff
@@ -318,3 +318,176 @@ def populate_impression_table():
...
+        result = self.client.post(
+            '/api/articles/no-article/comments/{}/thread/'.format(
+                self.create_comment.data['id']),
+            self.thread,
+            format="json"
+        )
+        self.assertEqual(result.status_code, status.HTTP_404_NOT_FOUND)
+        self.assertIn("The Article does not exist", str(result.data['detail']))
+
+    def test_tread_creation_with_non_existent_parent_comment(self):
```
Comment (kakaemma) [inline authors/apps/articles/tests.py]: Correct the spelling of _thread_

File: authors/apps/articles/views.py [hunk 8]
```diff
@@ -239,3 +244,156 @@ def delete(self, request, slug):
...
+
+            serializer = self.serializer_class(
+                data=serializer_data,
+                context=serializer_context
+            )
+            serializer.is_valid(raise_exception=True)
+            serializer.save()
+
+            comment = Comment()
+
```
Comment (kakaemma) [inline authors/apps/articles/views.py]: Remove this line

File: authors/apps/articles/views.py [hunk 9]
```diff
@@ -239,3 +244,156 @@ def delete(self, request, slug):
...
+            status=status.HTTP_200_OK
+        )
+
+    def destroy(self, instance, slug, id):
+        try:
+            comment = Comment.objects.get(id=id)
+            comment.delete()
+
+            comment_count = Comment()
+
```
Comment (kakaemma) [inline authors/apps/articles/views.py:329]: Remove this line

**Comment by RonKbS** [comment]
Please change the comment key-word to comments

**Comment by RonKbS** [comment]
Return a 404 if a user tries to retrieve a non-existent comment

**Comment by RonKbS** [comment]
A field should be added indicating if a thread exists on a comment.

File: authors/apps/articles/exceptions.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+from rest_framework.exceptions import APIException
+
+
+class NotFoundException(APIException):
+    """
+    Provides 404 not found error
+    """
+    default_code = "not_found"
+    default_detail = "The requested resource is not found."
```
Comment (RonKbS) [inline authors/apps/articles/exceptions.py]: Retain the tense in the message.

**Pull request merged**